ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForStructTest.java

```diff
@@ -0,0 +1,156 @@
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.parser.KsqlParser;
+import io.confluent.ksql.parser.SqlBaseParser.SingleStatementContext;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.CreateStreamAsSelect;
+import io.confluent.ksql.parser.tree.CreateTable;
+import io.confluent.ksql.parser.tree.CreateTableAsSelect;
+import io.confluent.ksql.parser.tree.DereferenceExpression;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.FunctionCall;
+import io.confluent.ksql.parser.tree.InsertInto;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.QuerySpecification;
+import io.confluent.ksql.parser.tree.SingleColumn;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.util.DataSourceExtractor;
+import io.confluent.ksql.util.MetaStoreFixture;
+import org.easymock.EasyMock;
+import org.junit.Before;
+import org.junit.Test;
+
+public class StatementRewriteForStructTest {
+
+  private static final KsqlParser KSQL_PARSER = new KsqlParser();
+
+  private MetaStore metaStore;
+
+  @Before
+  public void init() {
+    metaStore = MetaStoreFixture.getNewMetaStore(new TestFunctionRegistry());
+  }
+
+  @Test
+  public void shouldCreateCorrectFunctionCallExpression() {
+    final String simpleQuery = "SELECT iteminfo.category.name, address.state FROM orders;";
+    final Statement statement = KSQL_PARSER.buildAst(simpleQuery, metaStore).get(0);
+
+    final QuerySpecification querySpecification = getQuerySpecification(statement);
+    assertThat(querySpecification.getSelect().getSelectItems().size(), equalTo(2));
+    final Expression col0 = ((SingleColumn) querySpecification.getSelect().getSelectItems().get(0))
+        .getExpression();
+    final Expression col1 = ((SingleColumn) querySpecification.getSelect().getSelectItems().get(1))
+        .getExpression();
+
+    assertThat(col0, instanceOf(FunctionCall.class));
+    assertThat(col1, instanceOf(FunctionCall.class));
+
+    assertThat(col0.toString(), equalTo(
+        "FETCH_FIELD_FROM_STRUCT(FETCH_FIELD_FROM_STRUCT(ORDERS.ITEMINFO, 'CATEGORY'), 'NAME')"));
+    assertThat(col1.toString(), equalTo("FETCH_FIELD_FROM_STRUCT(ORDERS.ADDRESS, 'STATE')"));
+
+  }
+
+  @Test
+  public void shouldNotCreateFunctionCallIfNotNeeded() {
+    final String simpleQuery = "SELECT orderid FROM orders;";
+    final Statement statement = KSQL_PARSER.buildAst(simpleQuery, metaStore).get(0);
+
+    final QuerySpecification querySpecification = getQuerySpecification(statement);
+    assertThat(querySpecification.getSelect().getSelectItems().size(), equalTo(1));
+    final Expression col0 = ((SingleColumn) querySpecification.getSelect().getSelectItems().get(0))
+        .getExpression();
+
+    assertThat(col0, instanceOf(DereferenceExpression.class));
+    assertThat(col0.toString(), equalTo("ORDERS.ORDERID"));
+  }
+
```

Comment:
We should add a case where the array is itself inside a struct and contains struct elements that are dereferenced.

Reply:
We have such test cases in QueryTranslationTest.